<?xml version="1.0" ?>
<robot name="ttop" xmlns:xacro="http://ros.org/wiki/xacro">
  <joint name="torso_base_joint" type="fixed">
    <parent link="base_link"/>
    <child link="torso_base"/>
  </joint>
  <joint name="stewart_base_joint" type="revolute">
    <parent link="torso_base"/>
    <child link="stewart_base"/>
    <origin xyz="0     0     0.151" rpy="0 0 0"/>
    <axis xyz="0 0 1"/>
    <limit lower="0.0" upper="0.0" effort="0.0" velocity="0.1"/>
  </joint>
  <joint name="head_joint" type="floating">
    <parent link="stewart_base"/>
    <child link="head"/>
  </joint>
  <link name="base_link"/>
  <link name="torso_base">
    <inertial>
      <mass value="1000.0"/>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia ixx="1" ixy="0" ixz="0" iyy="1" iyz="0" izz="1"/>
    </inertial>
    <collision name="ttop__collision">
      <origin xyz="0     0     0" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.14" length="0.151"/>
      </geometry>
    </collision>
    <visual name="ttop__visual">
      <origin xyz="0     0     0" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.14" length="0.151"/>
      </geometry>
    </visual>
  </link>
  <link name="stewart_base">
    <inertial>
      <mass value="1.2"/>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia ixx="1" ixy="0" ixz="0" iyy="1" iyz="0" izz="1"/>
    </inertial>
    <collision name="ttop__collision">
      <!--<origin xyz="0     0    0.0375" rpy="0 0 0"/>-->
      <origin xyz="0     0    0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://ttop_sim/models/ttop/meshes/rotating_base_mesh.dae" scale="1.0 1.0 1.0"/>
      </geometry>
    </collision>
    <visual name="ttop__visual">
      <!--<origin xyz="0     0    0.0375" rpy="0 0 0"/>-->
      <origin xyz="0     0    0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://ttop_sim/models/ttop/meshes/rotating_base_mesh.dae" scale="1.0 1.0 1.0"/>
      </geometry>
    </visual>
  </link>
  <link name="head">
    <inertial>
      <mass value="0.001"/>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia ixx="0.001" ixy="0" ixz="0" iyy="0.001" iyz="0" izz="0.001"/>
    </inertial>
    <collision name="ttop__collision">
      <!--<origin xyz="0      0      0.228359" rpy="0       0      3.141593"/>-->
      <!--<origin xyz="0      0      0.379359" rpy="0       0      3.141593"/>-->
      <origin xyz="0      0      0" rpy="0       0      0"/>
      <geometry>
        <mesh filename="package://ttop_sim/models/ttop/meshes/stewart_platform_mesh.dae" scale="1.0 1.0 1.0"/>
      </geometry>
    </collision>
    <visual name="ttop__visual">
      <!--<origin xyz="0      0      0.228359" rpy="0       0      3.141593"/>-->
      <!--<origin xyz="0      0      0.379359" rpy="0       0      3.141593"/>-->
      <origin xyz="0      0      0" rpy="0       0      0"/>
      <geometry>
        <mesh filename="package://ttop_sim/models/ttop/meshes/stewart_platform_mesh.dae" scale="1.0 1.0 1.0"/>
      </geometry>
    </visual>
  </link>

  <xacro:macro name="sensor_d435" params="name:=camera">
    <xacro:property name="M_PI" value="3.1415926535897931" />

    <xacro:property name="d435_cam_depth_to_left_ir_offset" value="0.0"/>
    <xacro:property name="d435_cam_depth_to_right_ir_offset" value="-0.050"/>
    <xacro:property name="d435_cam_depth_to_color_offset" value="0.015"/>

    <xacro:property name="d435_cam_width" value="0.090"/>
    <xacro:property name="d435_cam_height" value="0.025"/>
    <xacro:property name="d435_cam_depth" value="0.02505"/>
    <xacro:property name="d435_cam_mount_from_center_offset" value="0.0149"/>
  

    <xacro:property name="d435_cam_depth_px" value="${d435_cam_mount_from_center_offset}"/>
    <xacro:property name="d435_cam_depth_py" value="0.0175"/>
    <xacro:property name="d435_cam_depth_pz" value="${d435_cam_height/2}"/>

    <material name="${name}_aluminum">
    <color rgba="0.5 0.5 0.5 1"/>
    </material>

    <joint name="${name}_link_joint" type="fixed">
      <parent link="head"/>
      <child link="${name}_link" />
    </joint> 

    <link name="${name}_link">
      <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <box size="${d435_cam_depth} ${d435_cam_width} ${d435_cam_height} "/>
        </geometry>
        <material name="${name}_aluminum"/>
      </visual>
      <collision>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
        <box size="${d435_cam_depth} ${d435_cam_width} ${d435_cam_height} "/>
        </geometry>
      </collision>
      <inertial>
        <mass value="0.564" />
        <origin xyz="0 0 0" />
        <inertia ixx="0.003881243" ixy="0.0" ixz="0.0" iyy="0.000498940" iyz="0.0" izz="0.003879257" />
      </inertial>
    </link>

    <joint name="${name}_depth_joint" type="fixed">
      <parent link="${name}_link"/>
      <child link="${name}_depth_frame" />
    </joint>
    <link name="${name}_depth_frame"/>

    <joint name="${name}_depth_optical_joint" type="fixed">
      <parent link="${name}_depth_frame" />
      <child link="${name}_depth_optical_frame" />
    </joint>
    <link name="${name}_depth_optical_frame"/>

    <joint name="${name}_color_joint" type="fixed">
      <parent link="${name}_depth_frame" />
      <child link="${name}_color_frame" />
    </joint>
    <link name="${name}_color_frame"/>

    <joint name="${name}_color_optical_joint" type="fixed">
      <parent link="${name}_color_frame" />
      <child link="${name}_color_optical_frame" />
    </joint>
    <link name="${name}_color_optical_frame"/>
  </xacro:macro>
</robot>
